purpleair-aqi.js

```diff
@@ -41,15 +42,93 @@ const SENSOR_ID = args.widgetParameter || "69223";
  * @property {string} lon
  */
 
+/**
+ * @typedef {object} LatLon
+ * @property {number} latitude
+ * @property {number} longitude
+ */
+
+/**
+ * Get the closest PurpleAir sensorId to the given location
+ *
+ * @param {LatLon} location
+ * @returns {Promise<number>}
+ */
+async function getSensorId({ latitude, longitude }) {
+  if (SENSOR_ID) return SENSOR_ID;
+
+  const BOUND_OFFSET = 0.05;
+
+  const nwLat = latitude + BOUND_OFFSET;
+  const seLat = latitude - BOUND_OFFSET;
+  const nwLng = longitude - BOUND_OFFSET;
+  const seLng = longitude + BOUND_OFFSET;
+
+  const req = new Request(
+    `${API_URL}/data.json?opt=1/mAQI/a10/cC5&fetch=true&nwlat=${nwLat}&selat=${seLat}&nwlng=${nwLng}&selng=${seLng}&fields=ID`
+  );
+
+  /** @type {{ code?: number; data?: Array<Array<number>>; fields?: Array<string>;  }} */
+  const res = await req.loadJSON();
+
+  const RATE_LIMIT = 429;
+  if (res.code === RATE_LIMIT) return DEFAULT_SENSOR_ID;
+
+  const { fields, data } = res;
+
+  const sensorIdIndex = fields.indexOf("ID");
+  const latIndex = fields.indexOf("Lat");
+  const lonIndex = fields.indexOf("Lon");
+  const typeIndex = fields.indexOf("Type");
+  const OUTDOOR = 0;
+
+  const [closestSensor] = data
+    .filter((datum) => datum[typeIndex] === OUTDOOR)
```

Comment:
If you don't want to exclude indoor sensors, remove this line (and lines 82-83).